src/psyclone/transformations.py

```diff
@@ -1252,6 +1251,148 @@ def apply(self, target_nodes, options=None):
+
+    @property
+    def omp_nowait(self):
+        ''' Returns whether or not this Single region uses a nowait
+            clause to remove the end barrier. The default value is False
+        '''
+        return self._omp_nowait
+
+    @property
+    def name(self):
+        '''
+        :returns: the name of this transformation as a string.
+        :rtype: str
+        '''
+        return "OMPSingleTrans"
+
+    @omp_nowait.setter
+    def omp_nowait(self, value):
+        ''' Sets the nowait property that will be specified by
+            this transformation. Checks that the value supplied in
+            :py:obj:`value` is a bool '''
+        if not isinstance(value, bool):
+            raise TransformationError("Expected nowait to be a bool "
+                                      "but got {0}".format(value))
+        self._omp_nowait = value
+
+    def _directive(self, children):
+        '''
+        Creates the type of directive needed for this sub-class of
+        transformation.
+        :param children: list of Nodes that will be the children of \
+                         the created directive.
+        :type children: list of :py:class:`psyclone.psyir.nodes.Node`
+        :rtype: :py:class:`psyclone.psyGen.OMPDoDirective`
+        :raises NotImplementedError: if a collapse argument is supplied
+        '''
+        _directive = nodes.OMPSingleDirective(children=children,
+                                        nowait=self.omp_nowait)
+        return _directive
+
+    def validate(self, node_list, options=None):
+        '''
+        Perform OpenMP-specific validation checks.
+
+        :param node_list: list of Nodes to put within serial region.
+        :type node_list: list of :py:class:`psyclone.psyir.nodes.Node`
+        :param options: a dictionary with options for transformations.
+        :type options: dictionary of string:values or None
+        :param bool options["node-type-check"]: this flag controls if the \
+                type of the nodes enclosed in the region should be tested \
+                to avoid using unsupported nodes inside a region.
+
+        :raises TransformationError: if the target Nodes are already within \
+                                     some OMP single region.
+        '''
+
+        if node_list[0].ancestor(nodes.OMPSingleDirective):
+            raise TransformationError("Error in OMPSingle transformation:" +
+                                      " cannot create an OpenMP SINGLE " +
+                                      "region within another OpenMP SINGLE region.")
+
+        # Now call the general validation checks
+        super(OMPSingleTrans, self).validate(node_list, options)
+
+    def apply(self, node, options=None):
+        '''Apply the OMPSingleTrans transformation to the specified node in a
+        Schedule. 
+
+        At code-generation time (when
+        :py:meth:`OMPLoopDirective.gen_code` is called), this node must be
```

Comment:
Depending on what you decide to do with `gen_code`, this should be extended or replaced with "or when the PSyIR tree is given to a backend")

Reply:
Done